# Generated by using Rcpp::compileAttributes() -> do not edit by hand
# Generator token: 10BE3573-1514-4C36-9D1C-5A225CD40393

#' Function to find the SNP data columns from an Impute 2 file
#'
#' Function to find the SNP data columns from an Impute 2 file.
#' Checks that the SNP name column exists and that there are no
#' duplicate columns.
#'
#' @param i2file
#' Name of impute 2 file
#' @param dataStart
#' Column where the genetic data starts. Only the columns before
#' the genetic data are checked for SNP data
#' @return
#' List containing status and an
#' Integer vector of length 5 will column numbers for
#' SNP name, chromosome, location in BP, reference allele
#' and alternate allele. These columns must have column
#' name 'SNP', 'CHR', 'BP', 'A1', 'A2', respectively
#' @export
FindImpute2SNPData <- function(i2file, dataStart) {
    .Call('_Impute2toBD_FindImpute2SNPData', PACKAGE = 'Impute2toBD', i2file, dataStart)
}

#' Function to read the subject IDs from an Impute 2 file
#'
#' Function to read the subject IDs from an Impute 2 file. This
#' requires that the file have a header
#'
#' @param i2file
#' Name of impute 2 file
#' @param dataStart
#' Column where the genetic data starts. This is also the column
#' where the subject IDs start.
#' @param usesFID
#' Indicator if family IDs are included in the header
#' @return
#' List containing status and a
#' data frame with the subject and family IDs. If no family ID is
#' in the file these values are set to blank, "", not missing.
#' @export
ReadImpute2Subjects <- function(i2file, dataStart, usesFID) {
    .Call('_Impute2toBD_ReadImpute2Subjects', PACKAGE = 'Impute2toBD', i2file, dataStart, usesFID)
}

#' Function to count the subject in an Impute 2 file
#'
#' Function to count the subject in an Impute 2 file. This
#' requires that the file not have a header. No error checking
#' of the genetic values are done.
#'
#' @param i2file
#' Name of impute 2 file
#' @param dataStart
#' Column where the genetic data starts. This is also the column
#' where the subject IDs start.
#' @param format
#' Indicator if family IDs are included in the header
#' @return
#' List containing status and the number of subjects found
#' @export
CountImpute2Subjects <- function(i2file, dataStart, format) {
    .Call('_Impute2toBD_CountImpute2Subjects', PACKAGE = 'Impute2toBD', i2file, dataStart, format)
}

#' Function to read an Impute 2 file
#'
#' Function to read an Impute 2 file
#'
#' @return
#' 0 success
#' 1 failure
#' @export
rcpp_hello <- function() {
    .Call('_Impute2toBD_rcpp_hello', PACKAGE = 'Impute2toBD')
}

#' Function to convert and Impute 2 file to binary dosage format
#'
#' Function to convert and Impute 2 file to binary dosage format
#'
#' @param impute2File
#' Name of Impute 2 file
#' @param header
#' Does the impute 2 file have a header
#' @param snpInfoCols
#' Column numbers for chromosome, SNP name, location in BP, allele 1 and allele 2, respecitively
#' @param dataStartCol
#' Column where dosage/genetic probabilities start
#' @param binaryDosageFile
#' Name of output file
#' @return
#' String reporting status
#' @export
Impute2toBD_C <- function(impute2File, header, snpInfoCols, dataStartCol, binaryDosageFile) {
    .Call('_Impute2toBD_Impute2toBD_C', PACKAGE = 'Impute2toBD', impute2File, header, snpInfoCols, dataStartCol, binaryDosageFile)
}

